libsolidity/inlineasm/AsmCodeGen.cpp

```diff
@@ -48,24 +50,61 @@ using namespace dev::solidity::assembly;
 
 struct GeneratorState
 {
-	GeneratorState(ErrorList& _errors, AsmAnalysisInfo& _analysisInfo, eth::Assembly& _assembly):
-		errors(_errors), info(_analysisInfo), assembly(_assembly) {}
+	GeneratorState(ErrorList& _errors, AsmAnalysisInfo& _analysisInfo):
+		errors(_errors), info(_analysisInfo) {}
+
+	ErrorList& errors;
+	AsmAnalysisInfo info;
+};
 
-	size_t newLabelId()
+class EthAssemblyAdapter: public julia::AbstractAssembly
+{
+public:
+	EthAssemblyAdapter(eth::Assembly& _assembly):
+		m_assembly(_assembly)
+	{
+	}
+	virtual void setSourceLocation(SourceLocation const& _location) override
+	{
+		m_assembly.setSourceLocation(_location);
+	}
+	virtual int stackHeight() const override { return m_assembly.deposit(); }
+	virtual void appendInstruction(solidity::Instruction _instruction) override
+	{
+		m_assembly.append(_instruction);
+	}
+	virtual void appendConstant(u256 const& _constant) override
+	{
+		m_assembly.append(_constant);
+	}
+	/// Append a label.
+	virtual void appendLabel(size_t _labelId) override
+	{
+		m_assembly.append(eth::AssemblyItem(eth::Tag, _labelId));
+	}
+	/// Append a label reference.
+	virtual void appendLabelReference(size_t _labelId) override
+	{
+		m_assembly.append(eth::AssemblyItem(eth::PushTag, _labelId));
+	}
+	virtual size_t newLabelId() override
+	{
+		return assemblyTagToIdentifier(m_assembly.newTag());
+	}
+	virtual void appendLinkerSymbol(std::string const& _linkerSymbol) override
 	{
-		return assemblyTagToIdentifier(assembly.newTag());
+		m_assembly.appendLibraryAddress(_linkerSymbol);
 	}
 
-	size_t assemblyTagToIdentifier(eth::AssemblyItem const& _tag) const
+private:
+	size_t assemblyTagToIdentifier(eth::AssemblyItem const& _tag)
```

Comment:
Also could think about pulling this out at some point (since ContrcatCompiler uses similar code).